File: Part 2 - Regression/Section 6 - Polynomial Regression/regresionPolinomica.R
library(ggplot2)
# Regresion polinomica
dataset = read.csv("Position_Salaries.csv")
# eliminar la primera columna del dataset
dataset = dataset[,2:3]

# regresion lineal
lin_reg = lm(formula = Salary ~ .,
             data = dataset)

# regresion polinomica
# agregar otro nivel al cuadrado
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ .,
              data = dataset)

ggplot() + 
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             color = "red") +
  geom_line(aes(x= dataset$Level, y = predict(lin_reg, newdata = dataset)),
             color = "blue") +
  ggtitle("Prediccion lineal del sueldo en funcion del nivel del empleado") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")


ggplot() + 
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             color = "red") +
  geom_line(aes(x= dataset$Level, y = predict(poly_reg, newdata = dataset)),
            color = "blue") +
  ggtitle("Prediccion polinomial del sueldo en funcion del nivel del empleado") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")


# Visualizacion de los resultados del modelo polinomico
# crear un rango de valores de 0.1 en 0.1 de x
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)

ggplot() + 
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             color = "red") +
  geom_line(aes(x= x_grid, y = predict(poly_reg, newdata = data.frame(Level = x_grid,
                                                                      Level2 = x_grid^2,
                                                                      Level3 = x_grid^3,
                                                                      Level4 = x_grid^4))),
            color = "blue") +
  ggtitle("Prediccion polinomial del sueldo en funcion del nivel del empleado x_grid") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")

# Prediccion de nuevos resultados con regresion linea
y_pred_lin = predict(lin_reg, newdata = data.frame(Level = 6.5))


# prediccion de nuevos resultados con regresion polinomica
y_pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                     Level2 = 6.5^2,
                                                     Level3 = 6.5^3,
                                                     Level4 = 6.5^4))

File: Part 2 - Regression/Section 6 - Polynomial Regression/regresionPolinomica.py
# -*- coding: utf-8 -*-
"""
Created on Thu Nov 21 16:56:25 2019

@author: darub
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
# funciona para hacer entrenamiento lineal multilineal y polinomial LinearRegression
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
dataset = pd.read_csv("Position_Salaries.csv")

#Tomar los valores desde la columna 1 y sin la ultima
x = dataset.iloc[:,1:-1].values
# Tomar la ultima columna
y = dataset.iloc[:,-1].values

# no se divide en trenamiento y testeo por que se son pocos datos y se buscan los valores intermedios

# Ajustar la regresion lienal con el dataset
lin_reg = LinearRegression()
lin_reg.fit(x,y)
# predecir el valor de 11 cuanto deberian pagar
y_pred_lin = lin_reg.predict([[11]])

# Ajustar la regresion polinomica con el dataset, valor por defecto alcuadrado degree
poly_reg = PolynomialFeatures(degree=3)
# elevar los datos al cuadrado
x_poly = poly_reg.fit_transform(x)

#Entrenar el modelo con los datos elevandos al cuadrado
lin_reg_2 = LinearRegression()
lin_reg_2.fit(x_poly, y)
y_pred_lin_2 = lin_reg_2.predict([[1,1,11,121]])


# Visualizacion de los resultados del modelo lienal
plt.scatter(x,y, color="red")
plt.plot(x, lin_reg.predict(x), color="blue")
plt.title("Modelo de regresin lineal")
plt.xlabel("Posicion del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()

# Visualizacion de los resultados del modelo polinomico
# crear un rango de valores de 0.1 en 0.1 de x
x_grid = np.arange(min(x), max(x), 0.1)
# redimencionar para que el vector sea una columna
x_grid =  x_grid.reshape(len(x_grid),1)

plt.scatter(x,y, color="red")
plt.plot(x_grid, lin_reg_2.predict(poly_reg.fit_transform(x_grid)), color="blue")
plt.title("Modelo de regresin polinomial")
plt.xlabel("Posicion del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()



























